### ISSUE TITLE
Felix handles duplicate IPs (e.g. ECMP) incorrectly

### ISSUE THREAD

**Issue by Symmetric** [Metaswitch/calico#350]
While testing calico-docker for ECMP support, I added workloads on two nodes, both using the same IP. I then added them to the same profile. Upon deleting one of the endpoints, the ipset for the profile did not contain the shared IP.

@fasaxc mentioned that he wasn't 100% certain that the Felix code handled duplicates correctly, and this appears to show that there is indeed a problem.

I have diags bundles if you need them, but `calicoctl diags` didn't autogenerate a URL for me, so I'll not bother uploading them somewhere unless you want them.

**Comment by Symmetric** [comment]
@fasaxc - not sure if this is another symptom of the same problem, but when adding multiple endpoints with the same IP on different hosts, the second endpoint to be created doesn't get a route to its cali* veth interface. The cali* interface is created, but it's unroutable. (The existing '<IP> via <OtherNode> dev eth1' route still exists).

The corresponding felix.log entries:

```
2015-04-13 23:10:25,609 [DEBUG][32] calico.felix.endpoint 195: Endpoint updated: {'ipv4_gateway': u'172.17.8.101', 'name': u'cali1d428cf0e23', 'profile_id': u'PROF_B', 'state': u'active', 'ipv6_nets': [], 'host': u'core-01', 'mac': u'e2:53:f2:9e:3a:4e', 'ipv4_nets': 
[u'192.168.1.2/32'], 'ipv6_gateway': None, 'id': u'1d428cf0e23211e4aa5c0800272d7481'}
2015-04-13 23:10:25,610 [DEBUG][32] calico.felix.futils 90: Calling out to system : ['ip', '-6', 'route', 'list', 'dev', u'cali1d428cf0e23']
2015-04-13 23:10:25,614 [DEBUG][31] calico.felix.devices 66: Existing routes to cali1d428cf0e23 : 192.168.1.2  scope link ,
2015-04-13 23:10:25,617 [DEBUG][31] calico.felix.devices 83: Found existing IP addresses : set(['192.168.1.2'])
```

**Comment by fasaxc** [comment]
This pull request adds partial anycast support: https://github.com/Metaswitch/calico/pull/362

**Comment by neiljerram** [comment]
Is #362 the complete fix for this issue?

**Comment by fasaxc** [comment]
@neiljerram I think #362 turned out to be all the felix code changes required to support the most basic form of anycast to VMs on multiple compute hosts.  However, I think @Symmetric found some additional bird config was required?

(Previously, we were worried that we might need to change the route programming to the kernel too, in case the duplicate routes clobbered each other.)

There are certainly future enhancements that we'll want to make in order to make anycast more useful.  For example, applying a closeness metric to the routes.

Since #362 had reasonably substantial code-review markups, I think we have to slip it to 0.18.

**Comment by neiljerram** [comment]
Sounds like we can get this into 0.18, but someone needs to step up and own retesting and making sure that any non-Calico-repo bits are also in place.  @Symmetric can you do that?

**Comment by Symmetric** [comment]
@fasaxc there were bird config changes required, and those are on master of calico-docker (as of last week).

@neiljerram I'm heads-down on the Kubernetes demo for CoreOS fest, so I might not have bandwidth for testing in the next couple days.

Could be a good rampup task for @alexaltair or @robbrockbank; will keep you posted.

**Comment by neiljerram** [comment]
Anyone know where we are on this issue now?

**Comment by ed-harrison** [comment]
@robbrockbank @fasaxc Should this still be open, or is it resolved by #362?

**Comment by fasaxc** [comment]
Felix now functions correctly (without clobbering its data structures) and we program the routes into the kernel (again without clobbering).  The main Calico repo doesn't have the BIRD changes needed to make this all work though.

Suggest we close this one and then open an enhancement issue for the BIRD config and investigate what else we need to do to flesh this out.
